grpc/src/main/java/GrpcMatrixSerializer.java

```diff
@@ -0,0 +1,66 @@
+import com.graphhopper.routing.*;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+import router.RouterOuterClass.*;
+
+public class GrpcMatrixSerializer {
+
+    public static MatrixRouteReply serialize(GHMRequest request, GHMResponse response) {
+        if (request.getFailFast() && response.hasInvalidPoints()) {
+            MatrixErrors matrixErrors = new MatrixErrors();
+            matrixErrors.addInvalidFromPoints(response.getInvalidFromPoints());
+            matrixErrors.addInvalidToPoints(response.getInvalidToPoints());
+            throw new MatrixCalculationException(matrixErrors);
+        } else {
+            int from_len = request.getFromPoints().size();
+            int to_len = request.getToPoints().size();
+            List<List<Long>> timeList = new ArrayList(from_len);
+            List<Long> timeRow;
+            List<List<Long>> distanceList = new ArrayList(from_len);
+            List<Long> distanceRow;
+            Iterator<MatrixElement> iter = response.getMatrixElementIterator();
+            MatrixErrors matrixErrors = new MatrixErrors();
+            StringBuilder debugBuilder = new StringBuilder();
+            debugBuilder.append(response.getDebugInfo());
+
+            for(int fromIndex = 0; fromIndex < from_len; ++fromIndex) {
+                timeRow = new ArrayList(to_len);
+                timeList.add(timeRow);
+                distanceRow = new ArrayList(to_len);
+                distanceList.add(distanceRow);
+
+                for(int toIndex = 0; toIndex < to_len; ++toIndex) {
+                    if (!iter.hasNext()) {
+                        throw new IllegalStateException("Internal error, matrix dimensions should be " + from_len + "x" + to_len + ", but failed to retrieve element (" + fromIndex + ", " + toIndex + ")");
+                    }
+
+                    MatrixElement element = iter.next();
+                    if (!element.isConnected()) {
+                        matrixErrors.addDisconnectedPair(element.getFromIndex(), element.getToIndex());
+                    }
+
+                    if (request.getFailFast() && matrixErrors.hasDisconnectedPairs()) {
+                        throw new MatrixCalculationException(matrixErrors);
+                    }
+
+                    long time = element.getTime();
+                    timeRow.add(time == 9223372036854775807L ? null : Math.round((double)time / 1000.0D));
```

Comment:
While I'd generally agree with this change, this class was one that I actually had to copy+paste from the graphhopper matrix API codebase + modify slightly to work in the GRPC context. So, almost none of this class was written by me, and it simply matches what the GH codebase has in it (except for the return statement, where we return a custom proto instead of the standard json response that the http codebase returns). 

The extremely-specific values being checked against here I believe are static variables in the true codebase, but because the matrix API isn't a public package, I can't download actual sources for this code, just decompiled files (that will have literal values in place of these variables). @michaz might be able to comment better on what these values actually are/how these checks actually work in the matrix code, but all we really care about is that the behavior is identical to what we use today.

So, I can in theory mess with this class if we'd like, but I think my mindset was that it was better to leave as much code as possible untouched in this class, to keep this code from getting out-of-sync with GH's internal codebase. Happy to consider other solutions, though!

Reply:
Ah, yes, I understand. I'd suggest not being careful or compatible or general with this code, and just to write directly what we want to do here, always.

The "interface" to GraphHopper is now the "API"-call in the middle of the procedure. Though I can't _totally_ promise it will never change, there's no need to keep code _around_ it stable for that case. If it changes, we'll have to adapt this code.